clouds/bigquery/modules/sql/quadbin/QUADBIN_TOCHILDREN.sql

```diff
@@ -3,20 +3,43 @@
 ----------------------------
 
 CREATE OR REPLACE FUNCTION `@@BQ_DATASET@@.QUADBIN_TOCHILDREN`
-(quadbin INT64, resolution INT64)
+(cell INT64, children_resolution INT64)
 RETURNS ARRAY<INT64>
 AS ((
-    IF(resolution < 0 OR resolution > 26 OR resolution < ((quadbin >> 52) & 0x1F),
+    IF(children_resolution < 0 OR children_resolution > 26 OR children_resolution < ((children_resolution >> 52) & 0x1F),
         ERROR('Invalid resolution'),
         (
-            WITH __zxy AS (
-                SELECT `@@BQ_DATASET@@.QUADBIN_TOZXY`(quadbin) AS tile
+            WITH
+            __constants AS (
+              SELECT
+                ~(0x1F << 52) AS zoom_level_mask,
+                ((cell >> 52) & 0x1F) AS tile_z
+            ),
+            __block_constants AS (
+              SELECT
+                (children_resolution - tile_z) AS resolution_diff,
+                (1 << ((children_resolution - tile_z) << 1)) AS block_range,
+                (52 - (children_resolution << 1)) AS block_shift
+              FROM
+                  __constants
+            ),
+            __childbase_constants AS (
+              SELECT
+                ((cell & zoom_level_mask) | (children_resolution << 52)) & ~((block_range - 1) << block_shift)
+                    AS child_base
+              FROM
+                __block_constants,
+                __constants
             )
-            SELECT ARRAY_AGG(`@@BQ_DATASET@@.QUADBIN_FROMZXY`(resolution, xs, ys))
-            FROM
-                __zxy,
-                UNNEST(GENERATE_ARRAY(tile.x << (resolution - tile.z), ((tile.x + 1) << (resolution - tile.z)) - 1)) AS xs,
-                UNNEST(GENERATE_ARRAY(tile.y << (resolution - tile.z), ((tile.y + 1) << (resolution - tile.z)) - 1)) AS ys
+            SELECT
+              ARRAY(
+                SELECT
+                    child_base | (block << block_shift)
+                FROM
+                    __block_constants,
+                    __childbase_constants,
+                    UNNEST(GENERATE_ARRAY(0, block_range - 1)) AS block
```

Comment:
This is going to hit the limits, we should iterate the block in two dimensions:
```
  UNNEST(GENERATE_ARRAY(0, sqrt_block_range - 1)) AS u,
  UNNEST(GENERATE_ARRAY(0, sqrt_block_range - 1)) AS v
```

Reply:
confirmed the issue already at zoom level 14
`    ----------------------------- RES 14

      at Object.<anonymous> (clouds/bigquery/modules/test/quadbin/QUADBIN_TOCHILDREN.test.js:5:17)

 FAIL  test/quadbin/QUADBIN_TOCHILDREN.test.js (20.675 s)
  ✕ QUADBIN_TOCHILDREN benchmark (20431 ms)

  ● QUADBIN_TOCHILDREN benchmark

    GENERATE_ARRAY(0, 1048575, 1) produced too many elements
`

Reply:
Some benchmarching
Seems that the new implementation affect mainly postgres.
BTW seems also that the most of the time is spent generating and returning the final array that is the bottleneck in performance and resources.

BQ |   |  
-- | -- | --
RES | DEV | PROD
5 | 1155 | 1255
6 | 908 | 955
7 | 881 | 879
8 | 909 | 1056
9 | 879 | 917
10 | 939 | 906
11 | 1259 | 1379
12 | 3185 | 2879
13 | 7514 | 7524
  |   |  
  |   |  
Postgres |   |  
RES | DEV | PROD
5 | 121.875286102295 | 142.597675323486
6 | 117.948055267334 | 207.051038742065
7 | 117.958068847656 | 185.067892074585
8 | 118.127822875977 | 200.172901153564
9 | 235.250949859619 | 642.972946166992
10 | 240.173578262329 | 2249.86267089844
11 | 366.314172744751 | 6419.11768913269
12 | 297.591686248779 | 24088.1743431091
13 | 2954.9765586853 | 98065.9940242767
14 | 8035.71534156799 |  
15 | 31047.4951267242 |  

Executing the same queries directly in the server but removing array creation and just coulting the result give the following tims (for BQ)


RES | MANUAL
-- | --
5 | 261
6 | 334
7 | 329
8 | 299
9 | 313
10 | 532
11 | 311
12 | 362
13 | 421
14 | 412
15 | 731
16 | 1000
17 | 5000
18 | 17000
19 | 74000
20 | 345000
21 | 1154000